src/stan/services/util/read_dense_mass_matrix.hpp

```diff
@@ -0,0 +1,50 @@
+#ifndef STAN_SERVICES_UTIL_READ_DENSE_MASS_MATRIX_HPP
+#define STAN_SERVICES_UTIL_READ_DENSE_MASS_MATRIX_HPP
+
+#include <stan/callbacks/writer.hpp>
+#include <stan/io/var_context.hpp>
+#include <stan/math/prim/mat.hpp>
+#include <limits>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace services {
+    namespace util {
+
+      /**
+       * Extract dense mass matrix from a var_context object.
+       *
+       * @param[in] init_mass_matrix a var_context with array of initial values
+       * @param[in] num_params expected number of row, column elements
+       * @param[in,out] error_writer message writer
+       * @throws std::domain_error if cannot read the mass matrix
+       * @return mass_matrix
+       */
+      Eigen::MatrixXd
+      read_dense_mass_matrix(stan::io::var_context& init_mass_matrix,
+                                 size_t num_params,
+                                 stan::callbacks::writer& error_writer) {
+        Eigen::MatrixXd inv_mass_matrix;
+        try {
+          init_mass_matrix.validate_dims("read dense mass matrix",
+                           "mass_matrix", "matrix",
+                           init_mass_matrix.to_vec(num_params, num_params));
+          std::vector<double> dense_vals =
+            init_mass_matrix.vals_r("mass_matrix");
+          inv_mass_matrix =
+            stan::math::to_matrix(dense_vals, num_params, num_params);
+        } catch (const std::exception& e) {
+          error_writer("Cannot get mass matrix from input file");
```

Comment:
Can we have two try/catch blocks here and be more verbose with the error messages?  In particular, one block and error for seeing if the right objects is in the given `var_context` and a second one to see if the object has the right shape.  We want the user to quickly be able to identify the problem.